src/Features/Hud/RhythmGame.cpp

```diff
@@ -0,0 +1,132 @@
+#include "RhythmGame.hpp"
+
+#include "Event.hpp"
+#include "Variable.hpp"
+#include "Modules/Engine.hpp"
+#include "Modules/Client.hpp"
+#include "Modules/Surface.hpp"
+#include "Features/GroundFramesCounter.hpp"
+#include "Features/Session.hpp"
+#include "Modules/Scheme.hpp"
+#include "Modules/Console.hpp"
+
+Variable sar_rhythmgame("sar_rhythmgame", "0", "Show a HUD indicating your groundframes as rhythm game like popups.\n");
+
+bool RhythmGameHud::ShouldDraw() {
+	return sar_rhythmgame.GetBool();
+}
+
+int RhythmGameHud::GetGroundframes() {
+	return groundframes;
+}
+
+bool prevGrounded = false;
+void RhythmGameHud::HandleGroundframeLogic(int slot, bool grounded) {
+	if (!prevGrounded && grounded) {
+		rhythmGameHud.groundframes = 0;
+	} else if (grounded) {
+		rhythmGameHud.groundframes++;
+	}
+
+	prevGrounded = grounded;
+}
+
+void RhythmGameHud::Paint(int slot) {
+	int screenWidth, screenHeight;
+	engine->GetScreenSize(nullptr, screenWidth, screenHeight);
+	auto font = scheme->GetFontByID(49);
+	float fh = surface->GetFontHeight(font);
+
+	for (int i = 0; i < popups.size(); i++) {
+		RhythmGamePopup popup = popups[i];
+
+		int x = popup.x;
+		int y = popup.y;
+
+		Color popupColor;
+		std::string popupText;
+		if (popup.type == 0) {
+			popupColor = perfectColor;
+			popupColor.a = popup.alpha;
+
+			popupText = "PERFECT";
+		} else if (popup.type == 1) {
+			popupColor = goodColor;
+			popupColor.a = popup.alpha;
+
+			popupText = "GOOD";
+		} else if (popup.type == 2) {
+			popupColor = okColor;
+			popupColor.a = popup.alpha;
+
+			popupText = "OK";
+		} else if (popup.type == 3) {
+			popupColor = badColor;
+			popupColor.a = popup.alpha;
+
+			popupText = "BAD";
+		} else {
+			return;
+		}
+
+		surface->DrawTxt(font, x, y, popupColor, popupText.c_str());
+		if (popup.streak > 1 && popup.type == 0) {
+			std::string combo = "(" + std::to_string(popup.streak) + "x)";
+			surface->DrawTxt(font, x + (surface->GetFontLength(font, popupText.c_str()) / 2) - (surface->GetFontLength(font, combo.c_str()) / 2), y + fh, Color{232, 179, 45, static_cast<uint8_t>(popup.alpha)}, combo.c_str());
+		}
+
+		popups[i].lifetime--;
+		popups[i].alpha -= 255 / 180;
+		popups[i].y -= 2;
+
+		if (popups[i].lifetime <= 0) {
+			popups.erase(popups.begin() + i);
```

Comment:
I'm pretty sure you will miss the next popup because it will change the size of the vector. You have two options: Only increment the index i when this branch is not hit or decrement i after the erase call.

Reply:
i added the i-- after the erase call, good now?